### PR TITLE
NIAD-107: acceptance optional fields

### PR DESCRIPTION
**Pull request by grzegorz-robak**

### CODE REVIEW

File: src/intTest/resources/outbound_uat_data/acceptance/type3-transferin-all.edifact.dat [hunk 1]
```diff
@@ -21,8 +21,8 @@ PNA+PAT+N/10/10:OPI+++SU:ASQUITH+FO:ROGER+TI:LORD+MI:ADAM+FS:DANIEL'
 DTM+329:19460723:102'
 PDI+1'
 NAD+PAT++THE MANSION HOUSE:1 MANSION LANE:ST PAULS CRAY:ORPINGTON:KENT+++++BR12 7ET'
-NAD+PER++THE HAUNTED HOUSE:12 SHADOWY LANE:HAUNTED GLADE:HORSFORTH:LEEDS'
 S02+2'
 PNA+PER++++SU:PREVIOUS'
-UNT+21+00000005'
-UNZ+1+00000004'
+NAD+PER++THE HAUNTED HOUSE:12 SHADOWY LANE:HAUNTED GLADE:HORSFORTH:LEEDS'
```
Comment (mattd-kainos) [inline src/intTest/resources/outbound_uat_data/acceptance/type3-transferin-all.edifact.dat]: This was correct originally, previous address belongs in the first occurrence of S02

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonOldAddressMapper.java [hunk 2]
```diff
@@ -41,4 +41,10 @@ private String getAddressLineOrNull(List<StringType> addressLines, int index) {
         }
         return addressLines.get(index).toString();
     }
+
+    @Override
+    public boolean canMap(Parameters parameters) {
+        return ParametersExtension.extractPatient(parameters)
+            .getAddress().size() > 1;
```
Comment (mattd-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonOldAddressMapper.java:48]: What would happen for `getAddress().size() > 1`? I think there would data provided in FHIR request that is ignored by the adaptor. Need to reject the request with 400 if anything is missing.

Comment (grzegorz-robak) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonOldAddressMapper.java:48]: It's condition for optional segment. Why would we reject request without optional data?

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/FreeTextMapperTest.java [hunk 3]
```diff
@@ -24,6 +24,14 @@ void when_FreeTextValueExistsAndValueIsNotSet_Then_CanMap() {
         Parameters parameters = new Parameters()
             .addParameter(ParameterNames.FREE_TEXT, "");
 
+        assertThat(freeTextMapper.canMap(parameters)).isTrue();
```
Comment (mattd-kainos) [inline src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/FreeTextMapperTest.java]: I'd expect this to be false.

Also update method name: Then_CanNotMap? threw me off initially

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/FreeTextMapperTest.java [hunk 4]
```diff
@@ -24,6 +24,14 @@ void when_FreeTextValueExistsAndValueIsNotSet_Then_CanMap() {
...
+    }
+
+    @Test
+    void when_FreeTextValueExistsAndValueIsNull_Then_CanMap() {
```
Comment (mattd-kainos) [inline src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/FreeTextMapperTest.java]: and method name here

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/FreeTextMapperTest.java (hunk 3 above)
Comment (grzegorz-robak) [reply src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/FreeTextMapperTest.java]: I also thought like that. But `canMap()` should check only for existence of value without validating it as validation code is inside each Segment class. So if we would validate here then no error would come up later (segment would be ignored without info to end-user). For invalid data each Segment code will throw exception during mapping.

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfExitMapperTest.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    private static final LocalDate LOCAL_DATE = LocalDate.parse(DATE_STRING);
+
+    @Test
+    void When_MappingDateOfEntry_Then_ExpectCorrectResult() {
+        Parameters parameters = new Parameters();
+        parameters.addParameter()
+            .setName(ParameterNames.EXIT_DATE)
+            .setValue(new StringType(DATE_STRING));
+
+        var PersonDateOfExitMapper = new PersonDateOfExitMapper();
```
Comment (mattd-kainos) [inline src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfExitMapperTest.java]: lowercase p for var name?

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/FreeTextMapperTest.java (hunk 3 above)
Comment (mattd-kainos) [reply src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/FreeTextMapperTest.java]: That makes sense 👍

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfExitMapperTest.java (hunk 5 above)
Comment (mattd-kainos) [reply src/test/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonDateOfExitMapperTest.java]: and two more below

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonOldAddressMapper.java (hunk 2 above)
Comment (mattd-kainos) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/mapper/PersonOldAddressMapper.java:48]: I misunderstood, and a few lines above it even explains why this is 1
.skip(1) // previous address is always the second occurrence of address in the Patient resource

**Review by mattd-kainos — APPROVED** [review]

**Review by Alicja-Kainos — APPROVED** [review]

**Pull request merged**